src/librustc_mir/transform/inline.rs

```diff
@@ -94,33 +95,39 @@ impl Inliner<'tcx> {
 
         loop {
             local_change = false;
-            while let Some(callsite) = callsites.pop_front() {
+            'callsites: while let Some(callsite) = callsites.pop_front() {
                 debug!("checking whether to inline callsite {:?}", callsite);
                 if !self.tcx.is_mir_available(callsite.callee) {
                     debug!("checking whether to inline callsite {:?} - MIR unavailable", callsite);
                     continue;
                 }
 
-                let self_node_id = self.tcx.hir().as_local_node_id(self.source.def_id()).unwrap();
-                let callee_node_id = self.tcx.hir().as_local_node_id(callsite.callee);
-
-                let callee_body = if let Some(callee_node_id) = callee_node_id {
+                // This may result in a cycle since the callee MIR is from this crate
+                // and we be optimizing it right now in this very same query stack.
+                if callsite.callee.is_local() {
                     // Avoid a cycle here by only using `optimized_mir` only if we have
-                    // a lower node id than the callee. This ensures that the callee will
-                    // not inline us. This trick only works without incremental compilation.
-                    // So don't do it if that is enabled.
-                    if !self.tcx.dep_graph.is_fully_enabled()
-                        && self_node_id.as_u32() < callee_node_id.as_u32()
-                    {
-                        self.tcx.optimized_mir(callsite.callee)
-                    } else {
-                        continue;
+                    // no mutual recursion. This ensures that the callee will
```

Comment:
This is direct recursion right? (IE, this comment should move down to the next check below)

Reply:
Oops yes